js_modules/dagster-ui/packages/ui-core/src/assets/catalog/useAssetCatalogGroupAndSortBy.tsx

```diff
@@ -0,0 +1,421 @@
+import {COMMON_COLLATOR, assertUnreachable} from '../../app/Util';
+import {usePrefixedCacheKey} from '../../app/usePrefixedCacheKey';
+import {AssetHealthFragment} from '../../asset-data/types/AssetHealthDataProvider.types';
+import {tokenForAssetKey} from '../../asset-graph/Utils';
+import {AssetHealthStatus} from '../../graphql/types';
+import {useQueryAndLocalStoragePersistedState} from '../../hooks/useQueryAndLocalStoragePersistedState';
+import {DagsterTag} from '../../runs/RunTag';
+import {buildRepoPathForHuman} from '../../workspace/buildRepoAddress';
+import {statusToIconAndColor} from '../AssetHealthSummary';
+import {AssetTableFragment} from '../types/AssetTableFragment.types';
+
+const SORT_ITEMS = [
+  {
+    key: 'materialization_asc' as const,
+    text: 'Materialization (new to old)',
+  },
+  {
+    key: 'materialization_desc' as const,
+    text: 'Materialization (old to new)',
+  },
+  {
+    key: 'key_asc' as const,
+    text: 'Asset key (a to z)',
+  },
+  {
+    key: 'key_desc' as const,
+    text: 'Asset key (z to a)',
+  },
+];
+
+const ITEMS_BY_KEY = SORT_ITEMS.reduce(
+  (acc, item) => {
+    acc[item.key] = item;
+    return acc;
+  },
+  {} as Record<(typeof SORT_ITEMS)[number]['key'], (typeof SORT_ITEMS)[number]>,
+);
+
+type UseAssetCatalogGroupAndSortByProps = {
+  liveDataByNode: Record<string, AssetHealthFragment>;
+  assetsByAssetKey: Map<string, AssetTableFragment>;
+};
+
+type SortBy = (typeof SORT_ITEMS)[number]['key'];
+
+export const useAssetCatalogGroupAndSortBy = ({
+  liveDataByNode,
+  assetsByAssetKey,
+}: UseAssetCatalogGroupAndSortByProps) => {
+  const [sortBy, setSortBy] = useQueryAndLocalStoragePersistedState<SortBy>({
+    localStorageKey: usePrefixedCacheKey('catalog-sortBy'),
+    isEmptyState: (state) => !state || state === 'materialization_asc',
+    decode: useCallback((json: ParsedQs) => {
+      if (
+        ['materialization_asc', 'materialization_desc', 'key_asc', 'key_desc'].includes(
+          json.sortBy as SortBy,
+        )
+      ) {
+        return json.sortBy as SortBy;
+      }
+      return 'materialization_asc';
+    }, []),
+    encode: useCallback((sortBy: SortBy) => ({sortBy}), []),
+  });
+
+  const [groupBy, setGroupBy] = useQueryAndLocalStoragePersistedState<AssetHealthGroupBy>({
+    localStorageKey: usePrefixedCacheKey('catalog-groupBy'),
+    isEmptyState: (state) => !state || state === AssetHealthGroupBy.health_status,
+    decode: useCallback((qs: ParsedQs) => {
+      if (qs.groupBy && GROUP_BY.includes(qs.groupBy as AssetHealthGroupBy)) {
+        return qs.groupBy as AssetHealthGroupBy;
+      }
+      return AssetHealthGroupBy.health_status;
+    }, []),
+    encode: useCallback((b: AssetHealthGroupBy) => ({groupBy: b}), []),
+  });
+
+  const grouped: Record<string, Grouped<any, any>> = useMemo(() => {
+    switch (groupBy) {
+      case AssetHealthGroupBy.code_location:
+        return groupByAttribute({
+          liveDataByNode: assetsByAssetKey,
```

Comment:
This reads a bit oddly - up above on line 50 these are given two separate types, so I'm surprised this assignment works, maybe the liveDataByNode param is used for something more generic in this context?